File: app/src/main/java/ru/radiationx/anilibria/MainActivity.java
package ru.radiationx.anilibria;

import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.MenuItem;
import android.widget.Toast;

import com.ittianyu.bottomnavigationviewex.BottomNavigationViewEx;

import java.util.List;

import ru.radiationx.anilibria.ui.fragments.release.ReleaseFragment;
import ru.radiationx.anilibria.ui.fragments.releases.ReleasesFragment;
import ru.terrakok.cicerone.Navigator;
import ru.terrakok.cicerone.android.SupportFragmentNavigator;
import ru.terrakok.cicerone.commands.BackTo;
import ru.terrakok.cicerone.commands.Command;

public class MainActivity extends AppCompatActivity {
    private BottomNavigationViewEx bottomTabs;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        bottomTabs = findViewById(R.id.bottom_tabs);

        bottomTabs.setOnNavigationItemSelectedListener(item -> {
            setTitle(item.getTitle());
            return true;
        });


        addMenuToBottom(R.string.fragment_title_releases, R.drawable.ic_releases)
                .setOnMenuItemClickListener(item -> {
                    navigator.applyCommand(new BackTo(Screens.RELEASES_LIST));
                    return false;
                });
        addMenuToBottom(R.string.fragment_title_news, R.drawable.ic_news)
                .setOnMenuItemClickListener(item -> {
                    navigator.applyCommand(new BackTo(Screens.RELEASES_LIST));
                    return false;
                });
        addMenuToBottom(R.string.fragment_title_videos, R.drawable.ic_videos)
                .setOnMenuItemClickListener(item -> {
                    navigator.applyCommand(new BackTo(Screens.RELEASES_LIST));
                    return false;
                });
        addMenuToBottom(R.string.fragment_title_blogs, R.drawable.ic_blogs)
                .setOnMenuItemClickListener(item -> {
                    navigator.applyCommand(new BackTo(Screens.RELEASES_LIST));
                    return false;
                });
        addMenuToBottom(R.string.fragment_title_other, R.drawable.ic_other)
                .setOnMenuItemClickListener(item -> {
                    navigator.applyCommand(new BackTo(Screens.RELEASES_LIST));
                    return false;
                });

        bottomTabs.enableItemShiftingMode(false);
        bottomTabs.enableShiftingMode(false);
        bottomTabs.setTextVisibility(false);
        bottomTabs.enableAnimation(false);

        Log.e("SUKA", "" + getSupportFragmentManager().getFragments().size());
        List<Fragment> fragments = getSupportFragmentManager().getFragments();
        for (Fragment fragment : fragments) {
            Log.e("SUKA", "Old fragments: " + fragment);
        }
        if (savedInstanceState == null) {
            App.get().getRouter().newRootScreen(Screens.RELEASES_LIST);
        }
    }

    private MenuItem addMenuToBottom(String title, @DrawableRes int iconId) {
        return bottomTabs.getMenu().add(title).setIcon(iconId);
    }

    private MenuItem addMenuToBottom(@StringRes int titleId, @DrawableRes int iconId) {
        return bottomTabs.getMenu().add(titleId).setIcon(iconId);
    }

    private void setTitle(String title) {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(title);
        }
    }

    private MenuItem getCurrentTab() {
        return bottomTabs.getMenu().getItem(bottomTabs.getCurrentItem());
    }

    private Navigator navigator =
            new SupportFragmentNavigator(getSupportFragmentManager(), R.id.fragments_container) {

                @Override
                public void applyCommand(Command command) {
                    super.applyCommand(command);
                    MenuItem item = getCurrentTab();
                    setTitle(item.getTitle());
                }

                @Override
                protected Fragment createFragment(String screenKey, Object data) {
                    switch (screenKey) {
                        case Screens.RELEASE_DETAILS: {
                            ReleaseFragment f = new ReleaseFragment();
                            if (data instanceof Bundle) {
                                f.setArguments((Bundle) data);
                            }
                            return f;
                        }
                        case Screens.RELEASES_LIST: {
                            ReleasesFragment fragment = new ReleasesFragment();
                            return fragment;
                        }
                        default:
                            throw new RuntimeException("Unknown screen key: " + screenKey);
                    }
                }

                @Override
                protected void showSystemMessage(String message) {
                    Toast.makeText(MainActivity.this, message, Toast.LENGTH_SHORT).show();
                }

                @Override
                protected void exit() {
                    finish();
                }
            };


    @Override
    protected void onResume() {
        super.onResume();
        App.get().getNavigatorHolder().setNavigator(navigator);
    }

    @Override
    protected void onPause() {
        super.onPause();
        App.get().getNavigatorHolder().removeNavigator();
    }
}

File: app/src/main/java/ru/radiationx/anilibria/ui/activities/MainActivity.kt
package ru.radiationx.anilibria.ui.activities

import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*
import ru.radiationx.anilibria.App
import ru.radiationx.anilibria.R
import ru.radiationx.anilibria.Screens
import ru.radiationx.anilibria.ui.common.BackButtonListener
import ru.radiationx.anilibria.ui.common.RouterProvider
import ru.radiationx.anilibria.ui.fragments.TabFragment
import ru.terrakok.cicerone.Navigator
import ru.terrakok.cicerone.Router
import ru.terrakok.cicerone.commands.Back
import ru.terrakok.cicerone.commands.Command
import ru.terrakok.cicerone.commands.Replace
import ru.terrakok.cicerone.commands.SystemMessage


class MainActivity : AppCompatActivity(), RouterProvider {

    override val router: Router = App.navigation.root.router
    private val navigationHolder = App.navigation.root.holder

    private val tabs = arrayOf(
            Tab(R.string.fragment_title_releases, R.drawable.ic_releases, Screens.RELEASES_LIST),
            Tab(R.string.fragment_title_news, R.drawable.ic_news, Screens.ARTICLES_LIST),
            Tab(R.string.fragment_title_videos, R.drawable.ic_videos, Screens.VIDEOS_LIST),
            Tab(R.string.fragment_title_blogs, R.drawable.ic_blogs, Screens.BLOGS_LIST),
            Tab(R.string.fragment_title_other, R.drawable.ic_other, Screens.OTHER_LIST)
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initContainers()
        initBottomMenu()

        bottomTabs.enableItemShiftingMode(false)
        bottomTabs.enableShiftingMode(false)
        bottomTabs.setTextVisibility(false)
        bottomTabs.enableAnimation(false)

        if (savedInstanceState == null) {
            bottomTabs.currentItem = 0
            router.newRootScreen(Screens.RELEASES_LIST)
        }
    }

    override fun onResume() {
        super.onResume()
        navigationHolder.setNavigator(navigatorNew)
    }

    override fun onPause() {
        super.onPause()
        navigationHolder.removeNavigator()
    }

    override fun onBackPressed() {
        val fragment = supportFragmentManager.findFragmentById(R.id.root_container)
        if (fragment != null
                && fragment is BackButtonListener
                && (fragment as BackButtonListener).onBackPressed()) {
            return
        } else {
            router.exit()
        }
    }

    private fun initContainers() {
        val fm = supportFragmentManager
        val ta = fm.beginTransaction()
        tabs.forEach { tab ->
            var fragment: Fragment? = fm.findFragmentByTag(tab.screenKey)
            if (fragment == null) {
                fragment = TabFragment.newInstance(tab.screenKey)
                ta.add(R.id.root_container, fragment, tab.screenKey)
                        .detach(fragment)
            }
        }
        ta.commitNow()
    }

    private fun initBottomMenu() {
        tabs.forEachIndexed { _, tab ->
            bottomTabs.menu
                    .add(tab.title)
                    .setIcon(tab.icon)
                    .setOnMenuItemClickListener {
                        router.replaceScreen(tab.screenKey)
                        false
                    }
        }
    }

    private val navigatorNew = object : Navigator {
        override fun applyCommand(command: Command) {
            Log.e("SUKA", "ApplyCommand " + command)
            if (command is Back) {
                finish()
            } else if (command is SystemMessage) {
                Toast.makeText(this@MainActivity, command.message, Toast.LENGTH_SHORT).show()
            } else if (command is Replace) {
                Log.e("SUKA", "Replace " + command.screenKey)
                val fm = supportFragmentManager
                val ta = fm.beginTransaction()
                tabs.forEach {
                    val fragment = fm.findFragmentByTag(it.screenKey)
                    if (it.screenKey == command.screenKey) {
                        ta.attach(fragment)
                    } else {
                        ta.detach(fragment)
                    }
                }
                ta.commitNow()
            }
        }
    }

    class Tab(val title: Int, val icon: Int, val screenKey: String)
}
